### ISSUE TITLE
Inconsistent behaviour when rolling up timeseries with an offset

### ISSUE THREAD

**Issue by immerrr** [VictoriaMetrics/VictoriaMetrics#976]
**Describe the bug**
It would be great [upgrading](https://victoriametrics.github.io/#how-to-upgrade) to [the latest avaialble release](https://github.com/VictoriaMetrics/VictoriaMetrics/releases)
and verifying whether the bug is reproducible there.
It is also recommended reading [troubleshooting docs](https://victoriametrics.github.io/#troubleshooting).

**To Reproduce**
Steps to reproduce the behavior.

###  Start a fresh victoriametrics 1.50.1 single-instance service
```
            "Cmd": [
                "-retentionPeriod=120",
                "-selfScrapeInterval=15s",
                "-storageDataPath=/victoria-metrics-data"
            ],
            "Image": "victoriametrics/victoria-metrics:v1.50.1",
```

### Create the data in the said instance

<details> 
<summary>Request/response</summary>

```
{
  "request": {
    "method": "POST",
    "url": "http://172.17.0.1:8013/api/v1/import",
    "headers": {
      "User-Agent": "python-requests/2.25.0",
      "Accept-Encoding": "gzip, deflate",
      "Accept": "*/*",
      "Connection": "keep-alive",
      "Content-Length": "364"
    },
    "body": {
      "metric": {
        "__name__": "test_metric",
        "ns": "namespace"
      },
      "values": [
        1.0,
        10.0,
        100.0,
        1000.0,
        10000.0,
        100000.0,
        1000000.0,
        10000000.0
      ],
      "timestamps": [
        1546300800000,
        1546300800001,
        1546387199999,
        1546387200000,
        1546387200001,
        1546473599999,
        1546473600000,
        1546473600001
      ]
    }
  },
  "response": {
    "status_code": 204,
    "url": "http://172.17.0.1:8013/api/v1/import",
    "headers": {
      "Date": "Sun, 20 Dec 2020 14:03:11 GMT"
    },
    "lines": []
  }
}
```
</details>

It will create datapoints as follows:
```
            '2019-01-01T00:00:00.000Z': 1,
            '2019-01-01T00:00:00.001Z': 10,
            '2019-01-01T23:59:59.999Z': 100,

            '2019-01-02T00:00:00.000Z': 1000,
            '2019-01-02T00:00:00.001Z': 10000,
            '2019-01-02T23:59:59.999Z': 100000,

            '2019-01-03T00:00:00.000Z': 1000000,
            '2019-01-03T00:00:00.001Z': 10000000,
```

### Run aggregation that is supposed to roll up to the beginning of the interval

By default, VM, like Prometheus, will aggregate data to the end of the interval, i.e. data for `2019-01-02T00:00:00.000Z` timestamp will contain samples from `2019-01-01T00:00:00.001Z` to `2019-01-02T00:00:00.000Z`. By using `offset -86399999ms` that can be changed, but in some cases this produces incorrect results, when samples at `2019-01-01T00:00:00.001Z` and `2019-01-01T23:59:59.999Z` are still aggregated into `2019-01-02T00:00:00.000Z` bucket.

<details>

<summary>When requesting range from 2018-01-01T00:00:00Z to 2020-01-01T00:00:00Z </summary>

```
{
  "request": {
    "method": "GET",
    "url": "http://172.17.0.1:8013/api/v1/query_range?query=sum_over_time%28test_metric+offset+-86399999ms%29&start=1514764800.0&end=1577836800.0&step=1d",
    "headers": {
      "User-Agent": "python-requests/2.25.0",
      "Accept-Encoding": "gzip, deflate",
      "Accept": "*/*",
      "Connection": "keep-alive"
    }
  },
  "response": {
    "status_code": 200,
    "url": "http://172.17.0.1:8013/api/v1/query_range?query=sum_over_time%28test_metric+offset+-86399999ms%29&start=1514764800.0&end=1577836800.0&step=1d",
    "headers": {
      "Access-Control-Allow-Origin": "*",
      "Content-Encoding": "gzip",
      "Content-Type": "application/json; charset=utf-8",
      "Date": "Sun, 20 Dec 2020 14:14:44 GMT",
      "Transfer-Encoding": "chunked"
    },
    "json": {
      "status": "success",
      "data": {
        "resultType": "matrix",
        "result": [
          {
            "metric": {
              "ns": "namespace",
            },
            "values": [
              [
                1546300800,
                "1"
              ],
              [
                1546387200,
                "1110"
              ],
              [
                1546473600,
                "1110000"
              ],
              [
                1546560000,
                "10000000"
              ]
            ]
          }
        ]
      }
    }
  },
  "query_params": [
    [
      "query",
      "sum_over_time(test_metric offset -86399999ms)"
    ],
    [
      "start",
      "1514764800.0"
    ],
    [
      "end",
      "1577836800.0"
    ],
    [
      "step",
      "1d"
    ]
  ]
}
```

</details>

**Expected behavior**

`sum_over_time` returns `111` for Jan 01 (indicating three samples were used), `111000` for Jan 02 (next three samples), and `11000000` for Jan 03.

<details>

<summary> Same query, but for range from 2018-12-23T00:00:00Z to 2019-01-12T00:00:00Z </summary>

```
{
  "request": {
    "method": "GET",
    "url": "http://172.17.0.1:8013/api/v1/query_range?query=sum_over_time%28test_metric+offset+-86399999ms%29&start=1545523200.0&end=1547251200.0&step=1d",
    "headers": {
      "User-Agent": "python-requests/2.25.0",
      "Accept-Encoding": "gzip, deflate",
      "Accept": "*/*",
      "Connection": "keep-alive"
    }
  },
  "response": {
    "status_code": 200,
    "url": "http://172.17.0.1:8013/api/v1/query_range?query=sum_over_time%28test_metric+offset+-86399999ms%29&start=1545523200.0&end=1547251200.0&step=1d",
    "headers": {
      "Access-Control-Allow-Origin": "*",
      "Content-Encoding": "gzip",
      "Content-Type": "application/json; charset=utf-8",
      "Date": "Sun, 20 Dec 2020 14:11:54 GMT",
      "Transfer-Encoding": "chunked"
    },
    "json": {
      "status": "success",
      "data": {
        "resultType": "matrix",
        "result": [
          {
            "metric": {
              "ns": "namespace",
            },
            "values": [
              [
                1546300800,
                "111"
              ],
              [
                1546387200,
                "111000"
              ],
              [
                1546473600,
                "11000000"
              ]
            ]
          }
        ]
      }
    }
  },
  "query_params": [
    [
      "query",
      "sum_over_time(test_metric offset -86399999ms)"
    ],
    [
      "start",
      "1545523200.0"
    ],
    [
      "end",
      "1547251200.0"
    ],
    [
      "step",
      "1d"
    ]
  ]
}
```

</details>

**Screenshots**

**Version**
The line returned when passing `--version` command line flag to binary. For example:
```
$ docker-compose run victoriametrics  -version
victoria-metrics-20201215-191549-tags-v1.50.1-0-g795e32be4
```

**Used command-line flags**

<details>

<summary> <code>curl -s 172.17.0.1:8013/metrics | grep ^flag<code> </summary>

```
$ curl -s 172.17.0.1:8013/metrics | grep ^flag
flag{name="bigMergeConcurrency", value="0"} 1
flag{name="csvTrimTimestamp", value="1ms"} 1
flag{name="dedup.minScrapeInterval", value="0s"} 1
flag{name="deleteAuthKey", value="secret"} 1
flag{name="denyQueriesOutsideRetention", value="false"} 1
flag{name="dryRun", value="false"} 1
flag{name="enableTCP6", value="false"} 1
flag{name="envflag.enable", value="false"} 1
flag{name="envflag.prefix", value=""} 1
flag{name="finalMergeDelay", value="30s"} 1
flag{name="forceFlushAuthKey", value="secret"} 1
flag{name="forceMergeAuthKey", value="secret"} 1
flag{name="fs.disableMmap", value="false"} 1
flag{name="graphiteListenAddr", value=""} 1
flag{name="graphiteTrimTimestamp", value="1s"} 1
flag{name="http.connTimeout", value="2m0s"} 1
flag{name="http.disableResponseCompression", value="false"} 1
flag{name="http.idleConnTimeout", value="1m0s"} 1
flag{name="http.maxGracefulShutdownDuration", value="7s"} 1
flag{name="http.pathPrefix", value=""} 1
flag{name="http.shutdownDelay", value="0s"} 1
flag{name="httpAuth.password", value="secret"} 1
flag{name="httpAuth.username", value=""} 1
flag{name="httpListenAddr", value=":8428"} 1
flag{name="import.maxLineLen", value="104857600"} 1
flag{name="influx.maxLineSize", value="262144"} 1
flag{name="influxListenAddr", value=""} 1
flag{name="influxMeasurementFieldSeparator", value="_"} 1
flag{name="influxSkipMeasurement", value="false"} 1
flag{name="influxSkipSingleField", value="false"} 1
flag{name="influxTrimTimestamp", value="1ms"} 1
flag{name="insert.maxQueueDuration", value="1m0s"} 1
flag{name="loggerDisableTimestamps", value="false"} 1
flag{name="loggerErrorsPerSecondLimit", value="0"} 1
flag{name="loggerFormat", value="default"} 1
flag{name="loggerLevel", value="INFO"} 1
flag{name="loggerOutput", value="stderr"} 1
flag{name="loggerWarnsPerSecondLimit", value="0"} 1
flag{name="maxConcurrentInserts", value="32"} 1
flag{name="maxInsertRequestSize", value="33554432"} 1
flag{name="maxLabelsPerTimeseries", value="30"} 1
flag{name="memory.allowedBytes", value="0"} 1
flag{name="memory.allowedPercent", value="60"} 1
flag{name="metricsAuthKey", value="secret"} 1
flag{name="opentsdbHTTPListenAddr", value=""} 1
flag{name="opentsdbListenAddr", value=""} 1
flag{name="opentsdbTrimTimestamp", value="1s"} 1
flag{name="opentsdbhttp.maxInsertRequestSize", value="33554432"} 1
flag{name="opentsdbhttpTrimTimestamp", value="1ms"} 1
flag{name="pprofAuthKey", value="secret"} 1
flag{name="precisionBits", value="64"} 1
flag{name="promscrape.config", value=""} 1
flag{name="promscrape.config.dryRun", value="false"} 1
flag{name="promscrape.config.strictParse", value="false"} 1
flag{name="promscrape.configCheckInterval", value="0s"} 1
flag{name="promscrape.consulSDCheckInterval", value="30s"} 1
flag{name="promscrape.disableCompression", value="false"} 1
flag{name="promscrape.disableKeepAlive", value="false"} 1
flag{name="promscrape.discovery.concurrency", value="100"} 1
flag{name="promscrape.discovery.concurrentWaitTime", value="1m0s"} 1
flag{name="promscrape.dnsSDCheckInterval", value="30s"} 1
flag{name="promscrape.dockerswarmSDCheckInterval", value="30s"} 1
flag{name="promscrape.dropOriginalLabels", value="false"} 1
flag{name="promscrape.ec2SDCheckInterval", value="1m0s"} 1
flag{name="promscrape.eurekaSDCheckInterval", value="30s"} 1
flag{name="promscrape.fileSDCheckInterval", value="30s"} 1
flag{name="promscrape.gceSDCheckInterval", value="1m0s"} 1
flag{name="promscrape.kubernetesSDCheckInterval", value="30s"} 1
flag{name="promscrape.maxDroppedTargets", value="1000"} 1
flag{name="promscrape.maxScrapeSize", value="16777216"} 1
flag{name="promscrape.openstackSDCheckInterval", value="30s"} 1
flag{name="promscrape.streamParse", value="false"} 1
flag{name="promscrape.suppressDuplicateScrapeTargetErrors", value="false"} 1
flag{name="promscrape.suppressScrapeErrors", value="false"} 1
flag{name="relabelConfig", value=""} 1
flag{name="retentionPeriod", value="120"} 1
flag{name="search.cacheTimestampOffset", value="5m0s"} 1
flag{name="search.disableCache", value="false"} 1
flag{name="search.latencyOffset", value="30s"} 1
flag{name="search.logSlowQueryDuration", value="5s"} 1
flag{name="search.maxConcurrentRequests", value="8"} 1
flag{name="search.maxExportDuration", value="720h0m0s"} 1
flag{name="search.maxLookback", value="0s"} 1
flag{name="search.maxPointsPerTimeseries", value="30000"} 1
flag{name="search.maxQueryDuration", value="30s"} 1
flag{name="search.maxQueryLen", value="16384"} 1
flag{name="search.maxQueueDuration", value="10s"} 1
flag{name="search.maxStalenessInterval", value="0s"} 1
flag{name="search.maxTagKeys", value="secret"} 1
flag{name="search.maxTagValueSuffixesPerSearch", value="100000"} 1
flag{name="search.maxTagValues", value="100000"} 1
flag{name="search.maxUniqueTimeseries", value="300000"} 1
flag{name="search.minStalenessInterval", value="0s"} 1
flag{name="search.resetCacheAuthKey", value="secret"} 1
flag{name="search.treatDotsAsIsInRegexps", value="false"} 1
flag{name="selfScrapeInstance", value="self"} 1
flag{name="selfScrapeInterval", value="15s"} 1
flag{name="selfScrapeJob", value="victoria-metrics"} 1
flag{name="smallMergeConcurrency", value="0"} 1
flag{name="snapshotAuthKey", value="secret"} 1
flag{name="storageDataPath", value="/victoria-metrics-data"} 1
flag{name="tls", value="false"} 1
flag{name="tlsCertFile", value=""} 1
flag{name="tlsKeyFile", value="secret"} 1
flag{name="version", value="false"} 1

```

</details>

**Comment by valyala** [comment]
VictoriaMetrics aligns timestamps for data points returned from `/api/v1/query_range` to multiple of `step` values if the number of returned data points exceeds 50. This allows using response cache when the graph is periodically refreshed on `last N hours/days/etc` time range. It looks this alignment results in unexpected responses in the case above. There are the following workarounds:
* To pass `nocache=1` query arg to `/api/v1/query_range`. This will disable response caching for the query and will remove timestamps alignment for returned data points.
* To pass `-search.disableCache` command-line flag to VictoriaMetrics (or to vmselect if you run [cluster version of VictoriaMetrics](https://victoriametrics.github.io/Cluster-VictoriaMetrics.html)).

**Comment by immerrr** [comment]
That's interesting information, but aren't the queries already aligned in the same way? Both of them start and end on midnight, and step is 24h.

**Comment by immerrr** [comment]
Let me check if nocache=1 helps here...

**Comment by valyala** [comment]
> That's interesting information, but aren't the queries already aligned in the same way? Both of them start and end on midnight, and step is 24h.

I also thought about this. It looks like there is some kind of [leap second](https://en.wikipedia.org/wiki/Leap_second) on the selected time range, which breaks the perfect per-day alginment.

**Comment by immerrr** [comment]
Hmm, indeed, `nocache=1` helps. I'll see if I can dive deep enough into the alignment code to see what is wrong there, because 
- epoch time does not observe leap seconds, i.e. all days contain 86400s and leap seconds are represented as the same timestamp repeating twice, so for an "external" observer it is transparent
- if the data is indeed realigned, i would expect the timestamps to be different in the response, but that is not the case

**Comment by immerrr** [comment]
Ok, the problem is still related to ocaching, but it is not related to leap seconds. 

During evaluation there is one more time when alignment is applied to separate expressions after offset:
https://github.com/immerrr/victoriametrics/blob/49e306ca9724dbd635023001ab0e72acabdce880/app/vmselect/promql/eval.go#L442-L447

In this case, the `-86399999ms` offset itself is the reason why the time series is not aligned, and why the result is different depending on whether or not caching is enabled.

So if I understand correctly, for bigger range queries VM may elect to essentially override `offset` values that come from queries if they are not a multiple of step. I think it would be more user friendly if VM instead of changing the outcome of the evaluation by returning a completely different timeseries would rather disable alignment/caching or take offset into account somehow (`metric+step=1d` is cache compatible with `metric offset 1d+step=1d` but not so much with `metric offset 1h+step=1d`). WDYT?

**Comment by immerrr** [comment]
I have found a curious hack in the code, related to the candlestick function. It seems to be tackling the same problem that I was trying to fix with `offset -86399999ms`: the fact that prometheus evaluation model looks back instead of looking forward when rolling up: https://github.com/VictoriaMetrics/VictoriaMetrics/issues/309. Maybe this can be somehow made into a reusable feature that is not bound to candlestick function itself?

**Comment by valyala** [comment]
@immerrr , thanks for further analysis of the code. It looks like the code you refer, which adjusts `offset` values for improved response caching, isn't needed. It has been removed in the commit c86286ec1d8b4e285304dfa54455274732ecede1 . Now the original query with `offset -86399999ms` should work as expected without the need to pass `nocache=1` query arg.

**Comment by valyala** [comment]
> I have found a curious hack in the code, related to the candlestick function. It seems to be tackling the same problem that I was trying to fix with offset -86399999ms: the fact that prometheus evaluation model looks back instead of looking forward when rolling up: #309. Maybe this can be somehow made into a reusable feature that is not bound to candlestick function itself?

This sounds interesting. I need to think about the best approach for providing this functionality.

**Comment by valyala** [comment]
The commit mentioned above has been included in VictoriaMetrics [v1.51.1](https://github.com/VictoriaMetrics/VictoriaMetrics/releases). Closing this issue as fixed.
